Support * Wildcard for quarkus.http.cors.origins

Postremus opened the pull request:
The star wildcard for the origins means, that any origin can access the service.
Previosly, this behavior could only be achieved by leaving the config property out completly.
Now you can either not set the origins config property, or use "*".

Adjust the documentation for * origins.
Fix naming of one of the tests.

fixes #5422

machi1990 commented on extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java:
```diff
@@ -18,7 +18,7 @@
 import io.quarkus.vertx.http.security.TestIdentityProvider;
 
-public class CoresSecurityTestCase {
+public class CORSSecurityTestCase {
```
Good catch.

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -90,7 +90,7 @@ public void handle(RoutingContext event) {
             }
 
-            boolean allowsOrigin = corsConfig.origins.isEmpty() || corsConfig.origins.contains(origin);
+            boolean allowsOrigin = corsConfig.origins.contains("*") || corsConfig.origins.contains(origin);
```
@Postremus Can you please  check it is a single value list too and use `get(0).equals` ? We probably should not allow some random lists which might have non wildcard and wildcard values :-)
There also should be a unit test, please copy and paste one of the tests there to test a `*`

sberyozkin commented on extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java:
```diff
@@ -18,7 +18,7 @@
 import io.quarkus.vertx.http.security.TestIdentityProvider;
 
-public class CoresSecurityTestCase {
+public class CORSSecurityTestCase {
```
@Postremus thanks for spotting the typo, it is a minor issue but lets keep the original naming pattern, `CorsSecurityTestCase`

Postremus replied:
@sberyozkin The other test classes in this package are also named CORSXXXXTestCase, which is why I went for this name.

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:
```diff
@@ -19,7 +19,7 @@
      * default: returns any requested origin as valid
      */
-    @ConfigItem
+    @ConfigItem(defaultValue = "*")
```
I'm not sure it should be a default value - that would mean all other properties should also be defaulted to `*`

machi1990 replied:
We should probably revert this and allowsAllOrigin if the list contains only `*` or `isEmpty()`

sberyozkin replied:
@machi1990 yes, I agree

Postremus replied:
We may need to allow `*` for more of the cors properties anyway in the future.
`Access-Control-Expose-Headers`, `Access-Control-Allow-Methods` and `Access-Control-Allow-Headers` allow * as wildcard value.

sberyozkin replied:
@Postremus sure, it makes sense for the consistency purposes. But lets keep the current pattern, it is either absent or set to `*`

sberyozkin commented on docs/src/main/asciidoc/http-reference.adoc:
```diff
@@ -127,7 +127,7 @@ following properties will be applied before passing the request on to its actual
 |===
 |Property Name|Default|Description
-|quarkus.http.cors.origins||The comma-separated list of origins allowed for CORS. The filter allows any origin if this is not
+|quarkus.http.cors.origins|*|The comma-separated list of origins allowed for CORS. The filter allows any origin if this is not
```
See my note about the default value. Perhaps it should read: `The filter allows any origin if this is not set or set to '*'' ?

sberyozkin commented:
@Postremus Hi, can you please do `mvn process-sources` in `quarkus-vertx-http` ?

As mentioned above, for the sake of consistency, all other CORS properties which were allowing for a wildcard when omitted should also be updated.

I have to admit I'm a bit not sure now what is the best approach, the one your originally started with (property default to `*`) or the one I proposed yesterday (keep the original pattern). @machi1990, should we ask @Postremus to continue with his original idea ? It seemed to me late yesterday that it was changing things, but now, I'm thinking, if a given property is omitted => it is an indirect `*` or it can be defaulted to `*` from the start, in which a given property is never omitted OOB, yet it does not change anything, it is still `*` by default :-).
CC @gsmet

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -90,7 +90,8 @@ public void handle(RoutingContext event) {
             }
 
-            boolean allowsOrigin = corsConfig.origins.isEmpty() || corsConfig.origins.contains(origin);
+            boolean isWildcardOrigin = corsConfig.origins.size() == 1 && "*".equals(corsConfig.origins.get(0));
```
Hmmm, isn't it a breaking change? If it's empty, it used to pass, now it doesn't. Is it really what we want? Or did I miss something.

Blocking merge until I got an answer to that.

sberyozkin replied:
Good catch, `corsConfig.origins.isEmpty()` should still be there.

sberyozkin commented:
@Postremus 
I'm sorry for my rather confusing review, I did it in a great rush. So, here is a proposal: please revert to your original idea by defaulting to `*` as it makes things simpler without breaking the compatibility, only enforce that a list containing a `*` is a single element list. And make sure that all other similar properties are supported consistently. Give it a go please when you get a chance

gsmet commented:
I would have liked to have this one in 1.0.0.Final. Can someone finish it before Friday end of day?

Postremus commented:
@sberyozkin sorry for not responding earlier.
@gsmet I will give it a go.

machi1990 commented:
> @sberyozkin sorry for not responding earlier.
> @gsmet I will give it a go.

Thanks @Postremus

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:
```diff
@@ -6,42 +6,37 @@
      */
-    @ConfigItem
-    public List<HttpMethod> methods;
+    public List<String> methods;
```
Why this change?

Postremus replied:
We can't use an Enum here, when we want to support * wildcard for methods as well.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -48,7 +47,7 @@ private void processRequestedHeaders(HttpServerResponse response, String allowHe
 
     private void processMethods(HttpServerResponse response, String allowMethodsValue) {
-        if (corsConfig.methods.isEmpty()) {
+        if (containsOnlyWildcard(corsConfig.headers)) {
```
I don't understand this change?

Postremus replied:
This change is done so that we also support * for methods as well. See new diff.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -90,29 +89,36 @@ public void handle(RoutingContext event) {
     }
+
+    private boolean containsOnlyWildcard(List<String> list) {
+        return list.size() == 1 && "*".equals(list.get(0));
```
I asked for it to support the empty list and I don't think you have addressed it.

Postremus replied:
done

gsmet reviewed (changes requested):
I added a few comments/questions, I don't think we will converge on that one today.

Let's target the next release.

gsmet commented:
Oh and I just stumbled upon the fact that this will conflict with https://github.com/quarkusio/quarkus/pull/5387/commits . We want to get this one in the tree ASAP as it's a very large one and it's a pain to rebase.

We will have to revisit that one after the merge.

machi1990 commented:
@Postremus I noticed a formatting issue in the CI, could you run `mvn formatter:format` in the `vertx-http` module? and force push your changes. Thanks.

Postremus commented:
@machi1990 done.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:
```diff
@@ -6,69 +6,58 @@
      */
-    @ConfigItem
+    @ConfigItem(defaultValue = "*")
     public Optional<List<String>> exposedHeaders;
```
@Postremus sorry I'm getting back to this one a bit late. I rebased and added a commit to improve the documentation formatting.

And while doing so, I saw this change. It looks like a significant change in behavior or am I wrong? Did it really expose all headers by default before?

gsmet replied:
@Postremus ping?

Postremus replied:
@gsmet pong :)
You are right. That is quite the change from previous behavior. I removed this change.

gsmet reviewed (changes requested):
I asked an additional question.

Postremus committed:
Support * Wildcard for quarkus.http.cors.origins
The star wildcard for the origins means, that any origin can access the service.
Previosly, this behavior could only be achieved by leaving the config property out completly.
Now you can either not set the origins config property, or use "*".

Adjust the documentation for * origins.
Add test case for this.

fixes #5422

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -27,7 +27,7 @@ public CORSFilter(CORSConfig corsConfig) {
 
     private void processRequestedHeaders(HttpServerResponse response, String allowHeadersValue) {
-        if (!corsConfig.headers.isPresent()) {
+        if (!corsConfig.headers.isPresent() || containsOnlyWildcard(corsConfig.headers)) {
```
You could remove the first part of the `if`.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -49,7 +49,7 @@ private void processRequestedHeaders(HttpServerResponse response, String allowHe
 
     private void processMethods(HttpServerResponse response, String allowMethodsValue) {
-        if (!corsConfig.methods.isPresent()) {
+        if (!corsConfig.methods.isPresent() || containsOnlyWildcard(corsConfig.methods)) {
```
You could remove the first part of the `if`.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -58,9 +58,10 @@ private void processMethods(HttpServerResponse response, String allowMethodsValu
-            for (HttpMethod configMethod : corsConfig.methods.get()) {
-                if (requestedMethods.contains(configMethod.name().toLowerCase())) {
-                    validRequestedMethods.add(configMethod.name());
+            for (String configMethod : corsConfig.methods.get()) {
```
I think you mistakenly reverted the config changes necessary for this to work?

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -91,30 +92,42 @@ public void handle(RoutingContext event) {
+            if (corsConfig.exposedHeaders.isPresent()) {
+                List<String> exposedHeaders = corsConfig.exposedHeaders.orElse(Collections.emptyList());
+                if (exposedHeaders.contains("*") && exposedHeaders.size() > 1) {
+                    exposedHeaders.remove("*");
```
I don't understand why this is necessary?

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -91,30 +92,42 @@ public void handle(RoutingContext event) {
 
-            final Optional<List<String>> exposedHeaders = corsConfig.exposedHeaders;
+            if (corsConfig.exposedHeaders.isPresent()) {
+                List<String> exposedHeaders = corsConfig.exposedHeaders.orElse(Collections.emptyList());
```
You already checked the value is present. No need for a `orElse` (I know it was preexisting but let's get rid of it).

gsmet reviewed (changes requested):
I added some comments inline.

I think you reverted some config changes that are needed. Unless I miss something.

sberyozkin commented:
Closing this one after #12142 has been merged, @Postremus thanks for starting this effort earlier
